### ISSUE TITLE
diagnostics/lsp: all buffers attached to client when underline = true

### ISSUE THREAD

**Issue by mjlbach** [neovim/neovim#15705]
### Neovim version (nvim -v)

0.6.0 commit ede5695eb194e9b607421415525177888b447753

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

Fedora 34

### Terminal name/version

kitty

### $TERM environment variable

xterm-kitty

### Installation

build from repo

### How to reproduce the issue

With the lspconfig minimal init.lua https://github.com/neovim/nvim-lspconfig/blob/master/test/minimal_init.lua

Add: 
```lua
    vim.lsp.handlers['textDocument/publishDiagnostics'] = vim.lsp.with(vim.lsp.diagnostic.on_publish_diagnostics, {
      underline = true,
    })
``` 
into on_attach

1. mkdir my_python_project
2. cd my_python_project && git init
3. `main.py`
```python
from . import test

def main():
    test.hello_world()
```
4. `test.py`
```python
def hello_world():
    print("hello world")

def hello_world_error(error):
    return x
```
```lua
    vim.lsp.handlers['textDocument/publishDiagnostics'] = vim.lsp.with(vim.lsp.diagnostic.on_publish_diagnostics, {
      underline = true,
    })
``` 
5. nvim main.py (only), wait about 15 seconds to ensure the server is started, call `:LspInfo`

6. All buffers are attached to the server (this is a huge issue with larger projects)

Note this can be fixed by:
```lua
    vim.lsp.handlers['textDocument/publishDiagnostics'] = vim.lsp.with(vim.lsp.diagnostic.on_publish_diagnostics, {
      underline = false,
      signs = false, -- this is required due to https://github.com/neovim/neovim/issues/15713
      virtual_text = false, -- this is required due to https://github.com/neovim/neovim/issues/15713

    })
``` 

### Expected behavior

Only user opened buffers are attached to the server.

### Actual behavior

All buffers in which there is a valid diagnostic are attached to the server.

### Notes,
Commenting out https://github.com/mjlbach/neovim/blob/340f77e78ef6589176822d53cf5507295bae771e/runtime/lua/vim/diagnostic.lua#L821-L827 fixes the issue in the case `{underline=true, virtual_text=false, signs=false}`

In 0.5 we don't call highlight.range on unloaded buffers (they have been bufadded ofc, and added to the diagnostic cache) until bufenter, now I'm seeing we do.

**Comment by fsouza** [comment]
@mjlbach ohhh this explains why neovim was going crazy at 100% CPU earlier today on a large project at work. I was planning to dig into it on Sunday, thanks for investigating it!

**Comment by mjlbach** [comment]
Closed by #15715
